File: source python/Organisms/Animals/Human.py
from Application.Organisms.Animals.Animal import Animal
from Application.Utilities import Colors
import pygame
import random


class Human(Animal):
    STRENGTH = 5
    INITIATIVE = 4
    ABILITY_FIRST_PHASE_TIME = 3
    ABILITY_SECOND_PHASE_TIME = 5
    ABILITY_MOVES = 2
    ABILITY_SECOND_PHASE_CHANCE = 50
    ABILITY_WAITING_TIME = 10

    def __init__(self, position, world, **kwargs):
        super().__init__(Human.STRENGTH, Human.INITIATIVE, position, world, **kwargs)
        if "loader" in kwargs:
            loader = kwargs["loader"]
            self.__timeSinceAbilityActivated = int(loader.Next())
        else:
            self.__timeSinceAbilityActivated = Human.ABILITY_WAITING_TIME + 1

    def Act(self):
        self.__timeSinceAbilityActivated += 1

        moves = 1
        if self.__timeSinceAbilityActivated < Human.ABILITY_FIRST_PHASE_TIME or (self.__timeSinceAbilityActivated < Human.ABILITY_SECOND_PHASE_TIME and random.randint(0, 99) < Human.ABILITY_SECOND_PHASE_CHANCE):
            moves = Human.ABILITY_MOVES

        print("Human is waiting for his turn")
        while moves != 0 and self.IsAlive():
            key = self.GetWorld().GetKey()

            if key == pygame.K_s and self.__ActivateSpecialAbility():
                moves += Human.ABILITY_MOVES - 1
            newPosition = self.GetPosition().GetMovement(key)
            if newPosition is None:
                continue
            if self.GetWorld().IsInBoundary(newPosition):
                self._GoTo(newPosition)
            moves -= 1

    def __ActivateSpecialAbility(self):
        if self.__timeSinceAbilityActivated >= Human.ABILITY_WAITING_TIME:
            self.__timeSinceAbilityActivated = 0
            return True

        return False

    @staticmethod
    def __str__():
        return "Human"

    def GetColor(self):
        return Colors.LIME

    def GetLetter(self):
        return "H"

    def Save(self, saver):
        super().Save(saver)
        saver.Insert(self.__timeSinceAbilityActivated)

File: source java/organism/animals/Human.java
package organism.animals;

import java.awt.Color;
import java.awt.event.KeyEvent;
import organism.CollisionResult;
import organism.Organism;
import organism.manager.Organisms;
import utilities.MyRandom;
import utilities.input.KeyPressedListener;
import utilities.input.Keyboard;
import utilities.position.Position;
import utilities.savingsystem.Loader;
import utilities.savingsystem.Saver;
import virtualworld.World;


public class Human extends Animal
{
    static final int STRENGTH = 5;
    static final int INITIATIVE = 4;
    
    static final int ABILITY_WAITING_TIME = 10;
    static final int ABILITY_FIRT_PHASE_TIME = 3;
    static final int ABILITY_SECOND_PHASE_TIME = 5;
    static final int ABILITY_MOVES = 2;
    static final int ABILITY_SECOND_PHASE_CHANCE = 50;
    
    static final int HUMAN_ABILITY_KEY = KeyEvent.VK_S;
    static final int MAX_KEY_PRESSED_COUNT = 3;
    static final int NO_KEY = -1000;
    
    private int timeSinceAbilityActivated_;
    private int keysPressed[] = {NO_KEY, NO_KEY, NO_KEY};
    private int keysPressedCount = 0;

    public Human(Position position, World world) 
    {
        super(STRENGTH, INITIATIVE, position, world);
        timeSinceAbilityActivated_ = ABILITY_WAITING_TIME + 1;
        AddKeyboardListener();
    }
    
    public Human(Loader loader, World world, Position position)
    {
        super(loader, world, position);
        timeSinceAbilityActivated_  = loader.NextInt();
        AddKeyboardListener();
    }
    
    private void AddKeyboardListener()
    {
        Keyboard.AddKeyPressedListener(new KeyPressedListener()
        {
            @Override
            public void keyPressed(KeyEvent e) 
            {
               
               if(keysPressedCount == MAX_KEY_PRESSED_COUNT)
               {
                   for(int i = 1; i<keysPressedCount; i++)
                       keysPressed[i - 1] = keysPressed[i];
                   keysPressed[MAX_KEY_PRESSED_COUNT - 1] = e.getKeyCode(); 
               }
               else
                   keysPressed[keysPressedCount++] = e.getKeyCode();
            }
            
        });
    }

    @Override
    public void Act()
    {
        IncreaseAbilityCounts();
        int moves = 1;
        if (timeSinceAbilityActivated_ < ABILITY_FIRT_PHASE_TIME || (timeSinceAbilityActivated_ < ABILITY_SECOND_PHASE_TIME && (new MyRandom()).Chance(ABILITY_SECOND_PHASE_CHANCE)))
		moves = ABILITY_MOVES;
        
        while(moves != 0 && isAlive())
        {
            int key = GetKey();
            if(key == NO_KEY)
                break;
            
            Position newPosition;
            if(key == HUMAN_ABILITY_KEY && ActivateSpecialAbility())
                moves += ABILITY_MOVES - 1;
            else
            {
                newPosition = GetPosition().GetMovement(key);
                if(newPosition == null)
                    break;

                if(GetWorld().IsInBoundary(newPosition))
                    GoTo(newPosition);
                
                moves--;
            }
        }
        
        keysPressedCount = 0;
    }
    
    @Override
    public CollisionResult Collide(Organism organism)
    {
        return CollisionResult.FIGHT;
    }
    
    @Override
    public void Save(Saver saver)
    {
        super.Save(saver);
        saver.Print(Integer.toString(timeSinceAbilityActivated_));
    }
    
    private boolean ActivateSpecialAbility()
    {
        if (timeSinceAbilityActivated_  >= ABILITY_WAITING_TIME)
	{
            timeSinceAbilityActivated_ = 0;
            GetWorld().AddToLogs(toString() + " activated special ability");
            return true;
	}
        GetWorld().AddToLogs(toString() + " can't activate special ability");
	return false;
    }
    
    private int GetKey()
    {
        if(keysPressedCount == 0)
            return NO_KEY;
        
        int ret = keysPressed[0];
        for(int i = 1; i<keysPressedCount; i++)
           keysPressed[i - 1] = keysPressed[i];
        return ret;
    }
    
    private void IncreaseAbilityCounts()
    {
        timeSinceAbilityActivated_ ++;
    }
    
    @Override
    public Organisms GetEnum() 
    {
        return Organisms.HUMAN;
    }
    
    @Override
    public Color GetColor()
    {
        return Color.ORANGE;
    }
    
    @Override
    public String toString()
    {
        return "Human";
    }

}
